shtdlib.sh

```diff
@@ -308,7 +308,7 @@ function shopt_decorator {
 # space separated string of the supported versions.
 function test_decorator {
     # If not running in a container
-    if [ "${FUNCNAME[0]}" != "${FUNCNAME[2]}" ] && ! grep -q docker /proc/1/cgroup ; then
+    if [ "${FUNCNAME[0]-unset}" != "${FUNCNAME[2]-unset}" ] && ! grep -q docker /proc/1/cgroup ; then
```

Comment:
FUNCNAME is an array containing the hierarchy of function calls leading to the current spot (a bit like a stack trace). This check just makes sure that when decorating another function we don't create an infinite loop by making sure the current function and the function before last aren't the same.

I can see cases where FUNCNAME2 is not set but if FUNCNAME0 is not set we're in real trouble.

I think the default values also need to be different and also "unset" is not a safe default since you might decorate a function called "unset". Perhaps a default value of "false" would be better since it's reserved and then only on FUNCNAME2 and we might want to consider doing that for the other decorators?

Also unset is a valid unix command ....

Reply:
Fixed